crates/net/src/http/response.rs

```diff
@@ -0,0 +1,268 @@
+        let mut body = vec![0; typed_buff.length() as usize];
+        typed_buff.copy_to(&mut body);
+        Ok(body)
+    }
+}
+
+impl From<web_sys::Response> for Response {
+    fn from(raw: web_sys::Response) -> Self {
+        Self(raw)
+    }
+}
+
+impl From<Response> for web_sys::Response {
+    fn from(res: Response) -> Self {
+        res.0
+    }
+}
+
+impl fmt::Debug for Response {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("Response")
+            .field("url", &self.url())
+            .field("redirected", &self.redirected())
+            .field("status", &self.status())
+            .field("headers", &self.headers())
+            .field("body_used", &self.body_used())
+            .finish()
+    }
+}
+
+/// A writable wrapper around `web_sys::Reponse`: an http response to be used with the `fetch` API
+/// on a server side javascript runtime
+pub struct ResponseBuilder {
+    headers: Headers,
+    options: web_sys::ResponseInit,
+    body: Option<ResponseBody>,
+}
+
+/// Possible initializers for request body
+#[derive(Debug)]
+pub enum ResponseBody {
+    /// Blob response body
+    Blob(web_sys::Blob),
+    /// Buffer response body
+    Buffer(js_sys::Object),
+    /// `Uint8Array` response body
+    U8(Vec<u8>),
+    /// `FormData` response body
+    Form(web_sys::FormData),
+    /// `URLSearchParams` response body
+    Search(web_sys::UrlSearchParams),
+    /// String response body
+    Str(String),
+    /// ReadableStream response body
+    Stream(web_sys::ReadableStream),
+}
+
+impl ResponseBuilder {
+    /// Creates a new response object
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    /// Replace _all_ the headers.
+    pub fn headers(mut self, headers: Headers) -> Self {
+        self.headers = headers;
+        self
+    }
+
+    /// Sets a header.
+    pub fn header(self, key: &str, value: &str) -> Self {
+        self.headers.set(key, value);
+        self
+    }
+
+    /// Set the status code
+    pub fn status(mut self, status: u16) -> Self {
+        self.options.status(status);
+        self
+    }
+
+    /// Set the status text
+    pub fn status_text(mut self, status_text: &str) -> Self {
+        self.options.status_text(status_text);
+        self
+    }
+
+    /// A convenience method to set JSON as response body
+    ///
+    /// # Note
+    ///
+    /// This method also sets the `Content-Type` header to `application/json`
+    #[cfg(feature = "json")]
+    #[cfg_attr(docsrs, doc(cfg(feature = "json")))]
+    pub fn json<T: serde::Serialize + ?Sized>(self, value: &T) -> Result<Response, Error> {
+        let json = serde_json::to_string(value)?;
+        self.header("Content-Type", "application/json")
+            .body(Some(ResponseBody::Str(json)))
+    }
+
+    /// Set the response body and return the response
+    pub fn body(mut self, data: Option<ResponseBody>) -> Result<Response, Error> {
```

Comment:
```suggestion
    pub fn body<T>(mut self, body: T) -> Result<Response, Error>
    where
        T: IntoRawResponse
    {

pub trait IntoRawResponse {
    fn into_raw_response(init: ResponseInit) -> web_sys::Response;
}
```

Instead of using an enum, I would suggest we allow any type that implements a certain trait (I have named it IntoRawResponse in this example) to be used as body.
This way we can avoid the full match below, which should also provide better dead code elimination.